### ISSUE TITLE
`jj file annotate` significantly slower than `git blame`

### ISSUE THREAD

**Issue by xdBronch** [martinvonz/jj#4878]
## Steps to Reproduce the Problem

take a reasonably large git repo and run `git blame`/`jj file annotate`, for example i used `lib/src/revset.rs` in the jj repo since its a fairly big file. jj is 5-10x slower

```
$ time git blame lib/src/revset.rs > /dev/null
real    0m 0.30s
user    0m 0.21s
sys     0m 0.09s
```
```
$ time jj file annotate lib/src/revset.rs > /dev/null
real    0m 1.60s
user    0m 1.38s
sys     0m 0.23s
```

## Expected Behavior
jj should perform better here

## Specifications

- Platform: x86-64 alpine linux 6.11.5
- Version: e4fdc91b690379716c73190647170f702684ee6e

**Comment by martinvonz** [comment]
Mabe it's the same root cause as #4674

**Comment by yuja** [comment]
> Unlike `jj log FILE`, `jj file annotate FILE` doesn't stream its output to stdout. Perhaps it could print the first line of the file once it determines the source change for the first line of the file (and for the second line, and so on)?

Technically it can, but I assume it wouldn't be worth the complexity. The first line is often old (e.g. shebang, copyright comment, etc.) Maybe we can instead add a flag to limit the search range or TUI that updates annotation incrementally.

**Comment by mernen** [comment]
While having an index would certainly be helpful, I suspect there are other problems lurking. For example, this should really be effectively instant in any repository, no matter how large:

    $ jj new && echo 'Line 1' > ANNOTATE-TEST && jj new && echo 'Line 2' >> ANNOTATE-TEST && jj st
    $ time jj --ignore-working-copy file annotate -T 'commit.commit_id() ++ "\n"' ANNOTATE-TEST

Only the two most recent commits should be traversed, but every commit in `..@` gets visited, so it scales with repo size. I tried to investigate, but this appears to be far too complex for someone who's new to the codebase and not familiar with how the graph traversal works; the furthest I could go is that the final call to [`RevsetGraphWalk::try_next()`](https://github.com/jj-vcs/jj/blob/c36bfafb6ac5f846ed4257e2dc4b41ff31a375b5/lib/src/default_index/revset_graph_iterator.rs#L320), after deciding on the `entry` to be returned, calls `remove_transitive_edges()`, which eventually ends up calling `consume_to()` which tests every commit with the revset filter (a fairly costly operation for `RevsetFilterPredicate::File`).

This appears to affect any file that contains at least one line that traces back to its first version (not unusual for a closing bracket or `import`/`use`/etc. or a copyright notice). If the file is reasonably recent (i.e. shouldn't need a deep traversal), the time added can be significant.

**Comment by WeetHet** [comment]
Yeah, the example in the issue isn't even close to the worst case. In JetBrains, our monorepo for IntelliJ has a couple million commits and `jj file annotate` takes about 100s compared to 1s `git blame`

Also, `jj debug index-changed-paths` sadly just fails:
```
$ jj debug index-changed-paths
Internal error: Failed to index commits at operation 09d1b326ebc33a13b2db3a3678a397bb5dcd7a15baca395341e66d013fae53fdd4ea993d4dfb8d52437d73d61692a635ed1c8d0703fc0fc91d15234a19b019c1
Caused by:
1: Invalid UTF-8 for object 094a54f7175239128a4d87d0a14605fa159c7b34 of type tree
2: invalid utf-8 sequence of 1 bytes from index 0
```

So I can't check if it would improve the situation
